### DISCUSSION TITLE
Identify Environment (Server or Client) For Better TypeScript Types

### DISCUSSION THREAD

**Discussion by LorenzoBloedow** [vercel/next.js#52354] [Ideas]
### Goals

The only goal of this request is to fix identify the environment the code is running on and with that information assign the appropriate types.

### Background

So recently I ran into a problem where I tried to use ``for await of`` (async iteration) from a readable stream generated in the server from a fetch request.

Unfortunately, TypeScript thought I was in a browser and so it wouldn't let me proceed without suppressing the following error: ```Type 'ReadableStream<Uint8Array>' is not an array type or a string type.```
This error happens because [the async iterable protocol for readable streams is currently only implemented in Firefox](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream#browser_compatibility), and the repository that generates the core TypeScript types states that [a feature has to have the support of at least two major browser engines to make it into TypeScript's core types](https://github.com/microsoft/TypeScript-DOM-lib-generator/blob/main/README.md#:~:text=A%20feature%20needs%20to%20be%20supported%20by%20two%20or%20more%20major%20browser%20engines%20to%20be%20included%20here).

However, I was executing this code in the server, where Next.js uses an extended implementation of fetch, which I've confirmed does support async iteration for readable streams.

The challenge is here is identifying the environment and then assigning the proper types according to said environment.

With the current way TypeScript is wired in Next, we're getting the wrong types, undermining the very problem that a type system is supposed to solve.

Maybe the async iteration is the only instance where there's a mismatch between types, or maybe (most likely) not.

Additionally, it's clear that, in the future, by implementing browser APIs on the server, more problems like this may arise.

If you know of another mismatch like this, I would love to hear about it.

### Proposal

Disclaimer: I'm by no means a TypeScript expert, so I have no idea if there's already a native solution to this problem baked within TypeScript, if so, please let me know down below.

My suggestion is simple on the surface but possibly hard to actually implement because it could mean forking TypeScript and I'm guessing no one wants that.

Regardless, I think we should use Next.js' algorithm for detecting whether a component is a client component or not, A.K.A, checking if the component itself or any of its parents has the ``"use client"`` directive at the top of the file.

Then, hint TypeScript about the appropriate types to use.

I'm happy to hear suggestions as to other ways this could be solved, or the best way to implement the solution above, and I'll gladly help contributing with a pull request if we ever get to that.

**Comment by icyJoseph** [comment]
How would you solve this on a Node.js env? Like, if I want to start a new TS project, what `tsconfig` and dependencies do I need t make it roll? 

AFAIK, fetch on Node TS is a bit of a touchy subject: https://github.com/DefinitelyTyped/DefinitelyTyped/issues/60924

Doing a `tsc --init` and adding `@types/node`:

```ts
import fs from "fs";

async function readData(url: string) {
  const response = await fetch(url);
  if (!response.body) return;

  for await (const chunk of response.body) {
    // Do something with each "chunk"
  }
  // Exit when done
}

``` 

This code complains about:

> Type 'ReadableStream<Uint8Array>' must have a '[Symbol.asyncIterator]()' method that returns an async iterator.ts(2504)

Even if I set target to "ESNEXT", it doesn't want to cooperate.

**Reply by LorenzoBloedow** [reply]
I'm not sure how to do it without type casting but you have to use the ``NodeJS.ReadableStream`` type by doing: ``(response.body as unknown as NodeJS.ReadableStream)``.

Currently there doesn't seem to be a way to switch contexts automatically, hence this discussion.

**Reply by icyJoseph** [reply]
Doesn't that mean we first have to solve this for `@types/node`? 

There's also the server-only package which can give some sort of safety. Perhaps within a server-only, poisoned module you could alias fetch and company (Response, Request) etc? 

Also, is the edge fetch aligned with native Node fetch? I'd guess so? 

This is a good initiative, though.

**Reply by LorenzoBloedow** [reply]
I don't think we have to touch ``@types/node`` at all since Next.js + TypeScript? (here's where I'm not so sure) would assign the types upon figuring out the environment, but yeah server-only modules would also be evaluated for their environment.

I'm not sure about fetch in the edge but it's good that you brought it up, it's something to consider.
